File: kardbrd_agent/worktree.py
"""Worktree manager for creating git worktrees per card."""

import logging
import subprocess
from pathlib import Path

logger = logging.getLogger("kardbrd_agent.worktree")


class WorktreeManager:
    """
    Manages git worktrees for kardbrd cards.

    Creates one worktree per card for isolated development.
    Worktrees are created as sibling directories to the base repo:
    ~/src/repo/ → ~/src/card-<short_id>/
    """

    def __init__(
        self,
        base_repo: Path,
        worktrees_dir: Path | None = None,
        setup_command: str | None = None,
        executor_type: str = "claude",
    ):
        """
        Initialize worktree manager.

        Args:
            base_repo: Path to the main git repository
            worktrees_dir: Optional directory for worktrees (defaults to base_repo parent)
            setup_command: Shell command to run in worktree after creation
                (e.g. "npm install", "uv sync"). None means skip setup.
            executor_type: The executor type ("claude" or "goose"). Controls
                which config symlinks are created.
        """
        self.base_repo = Path(base_repo).resolve()
        # Store worktrees in explicit dir or as sibling directories
        self.worktrees_base = (
            Path(worktrees_dir).resolve() if worktrees_dir else self.base_repo.parent
        )
        self.setup_command = setup_command
        self.executor_type = executor_type
        # Track active worktrees: card_id → worktree_path
        self.active_worktrees: dict[str, Path] = {}

    def _get_short_id(self, card_id: str) -> str:
        """Get truncated card ID (first 8 characters)."""
        return card_id[:8]

    def _get_worktree_path(self, card_id: str) -> Path:
        """Get the worktree path for a card (sibling directory)."""
        short_id = self._get_short_id(card_id)
        return self.worktrees_base / f"card-{short_id}"

    def _get_branch_name(self, card_id: str) -> str:
        """Get the branch name for a card."""
        short_id = self._get_short_id(card_id)
        return f"card/{short_id}"

    def _setup_symlinks(self, worktree_path: Path) -> None:
        """
        Set up symlinks for shared configuration files.

        Creates symlinks for .mcp.json, .env, and .claude/settings.local.json
        from the base repo to the worktree.
        """
        # Symlink shared configs
        for config_file in [".mcp.json", ".env"]:
            src = self.base_repo / config_file
            dst = worktree_path / config_file
            if src.exists() and not dst.exists():
                dst.symlink_to(src)
                logger.debug(f"Created symlink: {dst} -> {src}")

        # Claude settings (only for claude executor)
        if self.executor_type == "claude":
            claude_dir = worktree_path / ".claude"
            claude_dir.mkdir(exist_ok=True)
            settings_src = self.base_repo / ".claude" / "settings.local.json"
            settings_dst = claude_dir / "settings.local.json"
            if settings_src.exists() and not settings_dst.exists():
                settings_dst.symlink_to(settings_src)
                logger.debug(f"Created symlink: {settings_dst} -> {settings_src}")

    def _run_setup_command(self, worktree_path: Path) -> None:
        """
        Run the configured setup command in the worktree.

        Skips if no setup_command is configured.

        Args:
            worktree_path: Path to the worktree

        Raises:
            subprocess.CalledProcessError: If the setup command fails
        """
        if not self.setup_command:
            logger.debug(f"No setup command configured, skipping setup in {worktree_path}")
            return

        logger.info(f"Running setup command in {worktree_path}: {self.setup_command}")
        subprocess.run(
            self.setup_command,
            shell=True,
            cwd=worktree_path,
            check=True,
            capture_output=True,
        )

    def _update_main_branch(self) -> bool:
        """Fetch and fast-forward main before creating a new worktree."""
        try:
            # Fetch just main from origin
            subprocess.run(
                ["git", "fetch", "origin", "main"],
                cwd=self.base_repo,
                check=True,
                capture_output=True,
                text=True,
            )

            # Get current branch to restore later
            current = subprocess.run(
                ["git", "rev-parse", "--abbrev-ref", "HEAD"],
                cwd=self.base_repo,
                check=True,
                capture_output=True,
                text=True,
            ).stdout.strip()

            # Update main via fast-forward
            subprocess.run(
                ["git", "checkout", "main"],
                cwd=self.base_repo,
                check=True,
                capture_output=True,
                text=True,
            )
            subprocess.run(
                ["git", "pull", "--ff-only"],
                cwd=self.base_repo,
                check=True,
                capture_output=True,
                text=True,
            )

            # Restore previous branch if needed
            if current != "main":
                subprocess.run(
                    ["git", "checkout", current],
                    cwd=self.base_repo,
                    check=True,
                    capture_output=True,
                    text=True,
                )

            logger.info("Updated main branch from origin")
            return True
        except subprocess.CalledProcessError as e:
            logger.warning(f"Failed to update main branch: {e.stderr}")
            return False

    def create_worktree(self, card_id: str, branch_name: str | None = None) -> Path:
        """
        Create a worktree for a card.

        Creates worktree as a sibling directory: ~/src/repo/ → ~/src/card-<short_id>/
        Also sets up symlinks for shared configs and runs setup command if configured.

        Args:
            card_id: The card ID (used for worktree directory name)
            branch_name: Optional branch name (defaults to card/<short_id>)

        Returns:
            Path to the created worktree

        Raises:
            RuntimeError: If worktree creation fails
        """
        if not branch_name:
            branch_name = self._get_branch_name(card_id)

        worktree_path = self._get_worktree_path(card_id)

        # Check if worktree already exists
        if worktree_path.exists():
            logger.info(f"Worktree already exists: {worktree_path}")
            self.active_worktrees[card_id] = worktree_path
            return worktree_path

        logger.info(f"Creating worktree for card {card_id} at {worktree_path}")

        # Update main before creating worktree from latest
        self._update_main_branch()

        try:
            # Create worktree with new branch
            subprocess.run(
                ["git", "worktree", "add", "-b", branch_name, str(worktree_path)],
                cwd=self.base_repo,
                check=True,
                capture_output=True,
                text=True,
            )
            logger.info(f"Created worktree: {worktree_path} (branch: {branch_name})")

        except subprocess.CalledProcessError as e:
            # If branch already exists, try without -b
            if "already exists" in e.stderr.lower():
                logger.debug(f"Branch {branch_name} exists, creating worktree without -b")
                try:
                    subprocess.run(
                        ["git", "worktree", "add", str(worktree_path), branch_name],
                        cwd=self.base_repo,
                        check=True,
                        capture_output=True,
                        text=True,
                    )
                    logger.info(
                        f"Created worktree: {worktree_path} (existing branch: {branch_name})"
                    )
                except subprocess.CalledProcessError as e2:
                    raise RuntimeError(f"Failed to create worktree: {e2.stderr}") from e2
            else:
                raise RuntimeError(f"Failed to create worktree: {e.stderr}") from e

        # Setup symlinks for shared configs
        self._setup_symlinks(worktree_path)

        # Run setup command (e.g. "npm install", "uv sync") if configured
        self._run_setup_command(worktree_path)

        # Track active worktree
        self.active_worktrees[card_id] = worktree_path

        return worktree_path

    def remove_worktree(self, card_id: str, force: bool = False) -> None:
        """
        Remove a worktree for a card.

        Args:
            card_id: The card ID
            force: Force removal even if worktree has uncommitted changes

        Raises:
            RuntimeError: If worktree removal fails
        """
        worktree_path = self._get_worktree_path(card_id)

        # Remove from tracking
        self.active_worktrees.pop(card_id, None)

        if not worktree_path.exists():
            logger.debug(f"Worktree does not exist: {worktree_path}")
            return

        logger.info(f"Removing worktree for card {card_id}")

        try:
            cmd = ["git", "worktree", "remove", str(worktree_path)]
            if force:
                cmd.append("--force")

            subprocess.run(
                cmd,
                cwd=self.base_repo,
                check=True,
                capture_output=True,
                text=True,
            )
            logger.info(f"Removed worktree: {worktree_path}")

        except subprocess.CalledProcessError as e:
            raise RuntimeError(f"Failed to remove worktree: {e.stderr}") from e

    def get_worktree_path(self, card_id: str) -> Path | None:
        """
        Get the path to a worktree for a card.

        Args:
            card_id: The card ID

        Returns:
            Path to the worktree if it exists, None otherwise
        """
        worktree_path = self._get_worktree_path(card_id)
        return worktree_path if worktree_path.exists() else None

    def list_worktrees(self) -> list[tuple[str, Path]]:
        """
        List all active worktrees managed by this manager.

        Returns:
            List of (card_id, worktree_path) tuples
        """
        # Return tracked worktrees that still exist
        worktrees = []
        for card_id, path in self.active_worktrees.items():
            if path.exists() and (path / ".git").exists():
                worktrees.append((card_id, path))

        return worktrees

File: internal/worktree/worktree.go
package worktree

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type RunResult struct {
	Stdout string
	Stderr string
}

type Runner interface {
	Run(dir string, args []string) (RunResult, error)
}

type Manager struct {
	BaseRepo      string
	WorktreesBase string
	SetupCommand  string
	ExecutorType  string
	Runner        Runner

	active map[string]string
}

func NewManager(baseRepo string, worktreesDir string, setupCommand string, executorType string) *Manager {
	base := cleanAbs(baseRepo)
	worktreesBase := worktreesDir
	if worktreesBase == "" {
		worktreesBase = filepath.Dir(base)
	}
	if executorType == "" {
		executorType = "claude"
	}

	return &Manager{
		BaseRepo:      base,
		WorktreesBase: cleanAbs(worktreesBase),
		SetupCommand:  setupCommand,
		ExecutorType:  executorType,
		Runner:        commandRunner{},
		active:        map[string]string{},
	}
}

func (m *Manager) WorktreePath(cardID string) string {
	return filepath.Clean(filepath.Join(m.WorktreesBase, "card-"+shortID(cardID)))
}

func (m *Manager) BranchName(cardID string) string {
	return "card/" + shortID(cardID)
}

func (m *Manager) Create(cardID string, branchName string) (string, error) {
	if branchName == "" {
		branchName = m.BranchName(cardID)
	}
	path := m.WorktreePath(cardID)

	if exists(path) {
		m.active[cardID] = path
		return path, nil
	}
	if err := os.MkdirAll(m.WorktreesBase, 0o755); err != nil {
		return "", fmt.Errorf("create worktrees directory: %w", err)
	}

	_ = m.updateMainBranch()

	_, err := m.run("git", "worktree", "add", "-b", branchName, path)
	if err != nil {
		if !strings.Contains(strings.ToLower(err.Error()), "already exists") {
			return "", fmt.Errorf("failed to create worktree: %w", err)
		}
		if _, fallbackErr := m.run("git", "worktree", "add", path, branchName); fallbackErr != nil {
			return "", fmt.Errorf("failed to create worktree: %w", fallbackErr)
		}
	}

	if err := m.SetupSymlinks(path); err != nil {
		return "", err
	}
	if err := m.runSetupCommand(path); err != nil {
		return "", err
	}

	m.active[cardID] = path
	return path, nil
}

func (m *Manager) Remove(cardID string, force bool) error {
	path := m.WorktreePath(cardID)
	delete(m.active, cardID)

	if !exists(path) {
		return nil
	}

	args := []string{"git", "worktree", "remove", path}
	if force {
		args = append(args, "--force")
	}
	if _, err := m.run(args...); err != nil {
		return fmt.Errorf("failed to remove worktree: %w", err)
	}
	return nil
}

func (m *Manager) Get(cardID string) (string, bool) {
	path := m.WorktreePath(cardID)
	return path, exists(path)
}

func (m *Manager) ListActive() []string {
	var paths []string
	for _, path := range m.active {
		if exists(path) && exists(filepath.Join(path, ".git")) {
			paths = append(paths, path)
		}
	}
	return paths
}

func (m *Manager) SetupSymlinks(worktreePath string) error {
	if err := m.symlinkIfPresent(filepath.Join(m.BaseRepo, ".env"), filepath.Join(worktreePath, ".env")); err != nil {
		return err
	}

	switch strings.ToLower(m.ExecutorType) {
	case "claude":
		if err := os.MkdirAll(filepath.Join(worktreePath, ".claude"), 0o755); err != nil {
			return err
		}
		return m.symlinkIfPresent(
			filepath.Join(m.BaseRepo, ".claude", "settings.local.json"),
			filepath.Join(worktreePath, ".claude", "settings.local.json"),
		)
	case "codex":
		if err := m.symlinkDirIfPresent(filepath.Join(m.BaseRepo, ".agents", "skills"), filepath.Join(worktreePath, ".agents", "skills")); err != nil {
			return err
		}
		return m.symlinkDirIfPresent(filepath.Join(m.BaseRepo, ".codex", "skills"), filepath.Join(worktreePath, ".codex", "skills"))
	default:
		return nil
	}
}

func (m *Manager) updateMainBranch() error {
	if _, err := m.run("git", "fetch", "origin", "main"); err != nil {
		return err
	}
	result, err := m.run("git", "rev-parse", "--abbrev-ref", "HEAD")
	if err != nil {
		return err
	}
	current := strings.TrimSpace(result.Stdout)
	if _, err := m.run("git", "checkout", "main"); err != nil {
		return err
	}
	if _, err := m.run("git", "pull", "--ff-only"); err != nil {
		return err
	}
	if current != "" && current != "main" {
		if _, err := m.run("git", "checkout", current); err != nil {
			return err
		}
	}
	return nil
}

func (m *Manager) runSetupCommand(worktreePath string) error {
	if strings.TrimSpace(m.SetupCommand) == "" {
		return nil
	}
	if _, err := m.Runner.Run(worktreePath, []string{"sh", "-c", m.SetupCommand}); err != nil {
		return fmt.Errorf("setup command failed: %w", err)
	}
	return nil
}

func (m *Manager) symlinkIfPresent(src string, dst string) error {
	if !exists(src) || exists(dst) {
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	return os.Symlink(src, dst)
}

func (m *Manager) symlinkDirIfPresent(src string, dst string) error {
	info, err := os.Stat(src)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}
	if !info.IsDir() || exists(dst) {
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	return os.Symlink(src, dst)
}

func (m *Manager) run(args ...string) (RunResult, error) {
	return m.Runner.Run(m.BaseRepo, args)
}

type commandRunner struct{}

func (commandRunner) Run(dir string, args []string) (RunResult, error) {
	if len(args) == 0 {
		return RunResult{}, errors.New("missing command")
	}
	cmd := exec.Command(args[0], args[1:]...)
	cmd.Dir = dir

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	result := RunResult{Stdout: stdout.String(), Stderr: stderr.String()}
	if err != nil {
		return result, commandError{args: args, stderr: result.Stderr, err: err}
	}
	return result, nil
}

type commandError struct {
	args   []string
	stderr string
	err    error
}

func (e commandError) Error() string {
	if strings.TrimSpace(e.stderr) != "" {
		return strings.TrimSpace(e.stderr)
	}
	return fmt.Sprintf("%s: %v", strings.Join(e.args, " "), e.err)
}

func (e commandError) Unwrap() error {
	return e.err
}

func shortID(cardID string) string {
	if len(cardID) <= 8 {
		return cardID
	}
	return cardID[:8]
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func cleanAbs(path string) string {
	abs, err := filepath.Abs(path)
	if err != nil {
		return filepath.Clean(path)
	}
	return filepath.Clean(abs)
}
